File: sdks/node/src/runtime/stream-context.ts
/**
 * Stream Processing Context
 *
 * Wraps a StreamClient from the Stream SDK and manages the onActivate
 * callback lifecycle. Each embedded stream ID maps to one processing
 * context. The context owns the StreamClient instance and provides the
 * stream object interface to onActivate callbacks.
 *
 * onActivate runs as an async task (microtask/promise) on the stream
 * processing context, not the handler thread. Unhandled errors in
 * onActivate trigger automatic failStream.
 *
 * `StreamObject.end()` is task-scoped: it delegates to the agent-instance
 * `releaseStream(streamId, taskId)` hook, which evicts the per-task handle
 * cache entry, releases the registry taskId, and — only when the last
 * ref-holder releases — tears down the underlying StreamClient. On shared
 * streams the teardown never publishes a `stream_end` marker (see
 * SDK_CONTRACT §8.4.1a shared-affinity carve-out).
 */

import type { StreamClient, InboundMessage, StreamError } from '../stream/index.js';
import { log as baseLog } from './logger.js';

const log = (
  level: 'debug' | 'info' | 'warn' | 'error',
  message: string,
  meta?: Record<string, unknown>,
): void => baseLog('[StreamContext]', level, message, meta);

/** Hook the agent-instance passes to the stream wrapper so `end()` can
 * route through the registry + handle-cache eviction path. Decouples
 * stream-context.ts from agent-instance.ts (no import cycle). */
export interface AgentInstanceHooks {
  releaseStream: (streamId: string, taskId: string) => Promise<void>;
}

/** The stream object passed to onActivate and returned from createStream.
 *
 * The read/error/uuid surface mirrors the underlying `StreamClient` so
 * handlers and consumers see the same API. `onInboundDone` is intentionally
 * NOT forwarded — it's classified internal-only by SDK_CONTRACT §8.3.8.
 * Handlers needing a "stream drained" signal should `await` the
 * for-await-of loop on `bytes()` / `events()` / `inbound`.
 */
export interface StreamObject {
  readonly streamId: string;
  readonly channel: string;
  readonly isActive: boolean;
  readonly external: boolean;
  /** Underlying StreamClient uuid for log correlation. */
  readonly uuid: string;
  write(data: string | Uint8Array | unknown): void;
  end(): Promise<void>;
  /** Low-level wire iterator. Prefer `bytes()` / `events()` for decoded reads. */
  readonly inbound: AsyncIterable<InboundMessage>;
  /** Decoded byte iterator. Recommended for `format: bytes` streams. */
  bytes(): AsyncIterable<Uint8Array>;
  /** Flattened event iterator. Recommended for `format: events` streams. */
  events<T = unknown>(): AsyncIterable<T>;
  /** Node `Readable` adapter for piping into files / subprocesses. */
  readable(): Promise<import('node:stream').Readable>;
  onEnd(cb: () => void): void;
  /** Subscribe to stream-level errors (PAM revocation, network failure, …). */
  onError(cb: (err: StreamError) => void): void;
  /** T7a token (only available on external streams). */
  readonly token?: string;
  /** Activate an external stream (only available on external streams). */
  activate?(opts?: { metadata?: Record<string, unknown> }): Promise<void>;
}

export type OnActivateCallback = (stream: StreamObject) => void | Promise<void>;
export type FailStreamCallback = (streamId: string, error: string) => Promise<void>;

/**
 * Create a StreamObject wrapper around a StreamClient.
 * This is the shape passed to onActivate and returned from createStream.
 *
 * `hooks.releaseStream` — when provided, `end()` delegates to it so the
 * agent-instance can run task-scoped cleanup (registry release,
 * handle-cache eviction, affinity-gated teardown). Omit for synthetic
 * test wrappers with no agent-instance owner — `end()` falls back to
 * calling `client.end()` directly.
 */
export function createStreamObject(
  streamId: string,
  client: StreamClient,
  taskId?: string,
  hooks?: AgentInstanceHooks,
): StreamObject {
  return {
    get streamId() { return streamId; },
    get channel() { return client.channel; },
    get isActive() { return client.isActive; },
    get external() { return false; },
    get uuid() { return client.uuid; },
    write(data: string | Uint8Array | unknown) { client.write(data); },
    async end(): Promise<void> {
      if (hooks && taskId) {
        await hooks.releaseStream(streamId, taskId);
        return;
      }
      // Fallback for standalone usage (tests, non-agent-instance callers).
      await client.end();
    },
    get inbound() { return client.inbound; },
    bytes() { return client.bytes(); },
    events<T = unknown>() { return client.events<T>(); },
    readable() { return client.readable(); },
    onEnd(cb: () => void) { client.onEnd(cb); },
    onError(cb: (err: StreamError) => void) { client.onError(cb); },
  };
}

/**
 * Create a StreamObject for an external stream (no StreamClient).
 * write() and inbound throw; token and activate are available.
 */
export function createExternalStreamObject(
  streamId: string,
  channel: string,
  t7aToken: string,
  activateFn: (opts?: { metadata?: Record<string, unknown> }) => Promise<void>,
): StreamObject {
  let ended = false;
  const endCallbacks: Array<() => void> = [];

  return {
    get streamId() { return streamId; },
    get channel() { return channel; },
    get isActive() { return !ended; },
    get external() { return true; },
    get uuid(): string {
      throw new Error('Cannot read uuid on an external stream');
    },
    write() { throw new Error('Cannot write to an external stream -- use Stream SDK directly'); },
    async end() {
      ended = true;
      for (const cb of endCallbacks) { cb(); }
      endCallbacks.length = 0;
    },
    get inbound(): AsyncIterable<InboundMessage> {
      throw new Error('Cannot read from an external stream');
    },
    bytes(): AsyncIterable<Uint8Array> {
      throw new Error('Cannot read from an external stream');
    },
    events<T = unknown>(): AsyncIterable<T> {
      throw new Error('Cannot read from an external stream');
    },
    readable(): Promise<import('node:stream').Readable> {
      throw new Error('Cannot read from an external stream');
    },
    onEnd(cb: () => void) { endCallbacks.push(cb); },
    onError(_cb: (err: StreamError) => void) {
      throw new Error('Cannot subscribe to errors on an external stream');
    },
    get token() { return t7aToken; },
    activate: activateFn,
  };
}

/**
 * Run the onActivate callback on the stream processing context.
 * Returns a promise that resolves when the callback completes.
 * If the callback throws, the error is caught and failStream is called.
 */
export function runOnActivate(
  streamId: string,
  streamObject: StreamObject,
  callback: OnActivateCallback,
  failStreamCb: FailStreamCallback,
): Promise<void> {
  return Promise.resolve()
    .then(() => callback(streamObject))
    .catch(async (err: unknown) => {
      const errMsg = (err instanceof Error) ? err.message : String(err);
      log('error', `onActivate error for stream "${streamId}"`, {
        event: 'stream_context_on_activate_error',
        streamId,
        error: errMsg,
      });
      try {
        await failStreamCb(streamId, 'stream_crashed');
      } catch (failErr) {
        log('error', `failStream error for "${streamId}"`, {
          event: 'stream_context_fail_stream_error',
          streamId,
          error: failErr instanceof Error ? failErr.message : String(failErr),
        });
      }
    });
}

File: sdks/python/blocks_network/stream_context.py
"""
Stream Processing Context

Wraps a StreamClient from the Stream SDK and manages the on_activate
callback lifecycle. Each embedded stream ID maps to one processing
context. The context owns the StreamClient instance and provides the
stream object interface to on_activate callbacks.

Python-specific: on_activate runs on a dedicated daemon thread managed
by the stream context. The thread is started when the stream is
activated and stopped when the stream is destroyed.
"""

from __future__ import annotations

import logging
import threading
from typing import Any, Callable, Iterator, List, Optional

logger = logging.getLogger(__name__)

# Type alias for the on_activate callback signature
OnActivateCallback = Callable[["StreamObject"], None]
FailStreamCallback = Callable[[str, str], None]


class StreamObject:
    """The stream object passed to on_activate and returned from create_stream.

    Wraps a StreamClient from the Phase 2 Stream SDK and provides the
    developer-facing interface for stream I/O.

    ``end()`` delegates through ``release_stream`` (an ``AgentInstance``
    hook) so shared-affinity streams can be released task-scoped without
    tearing down the cross-task writer until the last ref-holder
    releases. See fix (d).
    """

    def __init__(
        self,
        stream_id: str,
        client: Any,
        *,
        external: bool = False,
        task_id: Optional[str] = None,
        release_stream: Optional[Callable[[str, str], None]] = None,
    ) -> None:
        self._stream_id = stream_id
        self._client = client
        self._external = external
        self._task_id = task_id
        self._release_stream = release_stream

    @property
    def stream_id(self) -> str:
        return self._stream_id

    @property
    def channel(self) -> str:
        return self._client.channel

    @property
    def is_active(self) -> bool:
        return self._client.is_active

    @property
    def external(self) -> bool:
        return self._external

    def write(self, data: Any) -> None:
        """Write data to the stream. Delegates to StreamClient.write()."""
        self._client.write(data)

    def end(self) -> None:
        """End the stream.

        When bound to an ``AgentInstance`` release hook (and a task id),
        delegates to the task-scoped release path:
        - Evicts the per-task shared-stream handle cache entry.
        - Decrements the registry's task tracking for the stream.
        - Shared streams: underlying ``StreamClient`` is kept alive
          until the last ref-holder releases, and no ``stream_end``
          marker is ever published on per-task cleanup.
        - Dedicated streams: the final release tears down the
          ``StreamClient`` which still publishes the end marker.

        Falls back to a direct ``StreamClient.end()`` (legacy / tests
        that construct a StreamObject without a hook).
        """
        if self._release_stream is not None and self._task_id is not None:
            self._release_stream(self._stream_id, self._task_id)
            return
        self._client.end()

    @property
    def inbound(self) -> Iterator:
        """Inbound message iterator. Delegates to StreamClient.inbound.

        Low-level wire iterator. For most read paths, prefer the decoded
        helpers ``bytes()`` (for ``format: bytes``) or ``events()`` (for
        ``format: events``).
        """
        return self._client.inbound

    @property
    def uuid(self) -> str:
        """Underlying StreamClient uuid for log correlation."""
        return self._client.uuid

    def bytes(self) -> Iterator[bytes]:
        """Decoded byte iterator. Delegates to StreamClient.bytes()."""
        return self._client.bytes()

    def events(self) -> Iterator[Any]:
        """Flattened event iterator. Delegates to StreamClient.events()."""
        return self._client.events()

    def as_file(self) -> Any:
        """File-like wrapper for ``shutil.copyfileobj`` / subprocess pipe.

        Delegates to ``StreamClient.as_file()``.
        """
        return self._client.as_file()

    def on_end(self, cb: Callable[[], None]) -> None:
        """Register a callback to fire when the stream ends."""
        self._client.on_end(cb)

    def on_error(self, cb: Callable[[Any], None]) -> None:
        """Register a callback to fire on stream-level errors.

        Delegates to ``StreamClient.on_error()``. The callback receives a
        :class:`StreamError` payload with the PubNub status category, the
        affected channel, a timestamp, and a ``fatal`` flag.

        Note: ``StreamClient.on_error`` only appends to its callback list;
        it does not buffer or replay past errors. Register the callback
        before the read path activates.
        """
        self._client.on_error(cb)


class ExternalStreamObject:
    """Stream object for external streams (no StreamClient).

    write() and inbound throw; token and activate are available.
    """

    def __init__(
        self,
        stream_id: str,
        channel: str,
        t7a_token: str,
        activate_fn: Callable,
    ) -> None:
        self._stream_id = stream_id
        self._channel = channel
        self._t7a_token = t7a_token
        self._activate_fn = activate_fn
        self._ended = False
        self._end_callbacks: List[Callable[[], None]] = []

    @property
    def stream_id(self) -> str:
        return self._stream_id

    @property
    def channel(self) -> str:
        return self._channel

    @property
    def is_active(self) -> bool:
        return not self._ended

    @property
    def external(self) -> bool:
        return True

    def write(self, data: Any) -> None:
        raise RuntimeError(
            "Cannot write to an external stream -- use Stream SDK directly"
        )

    def end(self) -> None:
        self._ended = True
        for cb in self._end_callbacks:
            cb()
        self._end_callbacks.clear()

    @property
    def inbound(self) -> Iterator:
        raise RuntimeError("Cannot read from an external stream")

    @property
    def uuid(self) -> str:
        raise RuntimeError("Cannot read uuid on an external stream")

    def bytes(self) -> Iterator[bytes]:
        raise RuntimeError("Cannot read from an external stream")

    def events(self) -> Iterator[Any]:
        raise RuntimeError("Cannot read from an external stream")

    def as_file(self) -> Any:
        raise RuntimeError("Cannot read from an external stream")

    def on_end(self, cb: Callable[[], None]) -> None:
        self._end_callbacks.append(cb)

    def on_error(self, cb: Callable[[Any], None]) -> None:
        raise RuntimeError("Cannot subscribe to errors on an external stream")

    @property
    def token(self) -> str:
        """T7a token from Phase 1 token_request handshake."""
        return self._t7a_token

    def activate(self, *, metadata: Optional[dict] = None) -> None:
        """Trigger Phase 2 activate handshake."""
        self._activate_fn(metadata=metadata)


def run_on_activate(
    stream_id: str,
    stream_object: Any,
    callback: OnActivateCallback,
    fail_stream_cb: FailStreamCallback,
) -> threading.Thread:
    """Run the on_activate callback on a dedicated daemon thread.

    Returns the thread object for tracking. If the callback throws,
    the error is caught and fail_stream is called automatically.
    """

    def _worker() -> None:
        try:
            callback(stream_object)
        except Exception as exc:
            err_msg = str(exc) if str(exc) else "stream_crashed"
            logger.error(
                "[StreamContext] on_activate error for stream %r: %s",
                stream_id,
                err_msg,
            )
            try:
                fail_stream_cb(stream_id, "stream_crashed")
            except Exception as fail_exc:
                logger.error(
                    "[StreamContext] fail_stream error for %r: %s",
                    stream_id,
                    fail_exc,
                )

    thread = threading.Thread(target=_worker, daemon=True)
    thread.start()
    return thread
